int32 algorithm
int32 state
---
bool success
string message
